src/workclaimer.js

```diff
@@ -0,0 +1,298 @@
+let events      = require('events');
+let taskcluster = require('taskcluster-client');
+
+/**
+ * HintPoller polls for hints for pending tasks.
+ *
+ * The azure queues don't know if a task is pending they just store hints of
+ * pending tasks. To be understood this way:
+ *  A) If a task is pending, there is a hint of the task in an azure queue,
+ *  B) If there is an hint in an azure queue, it may or may not be pending.
+ *
+ * It's an if, but not an only-if (think over-approximation).
+ */
+class HintPoller {
+  constructor(parent, provisionerId, workerType) {
+    this.parent = parent;
+    this.provisionerId = provisionerId;
+    this.workerType = workerType;
+    this.requests = [];
+    this.started = false;
+  }
+
+  requestClaim(count, aborted) {
+    // Make a request for count tasks
+    let request = null;
+    let result = new Promise((resolve, reject) => {
+      request = {resolve, reject, count};
+    });
+    this.requests.push(request);
+
+    aborted.then(() => {
+      // Remove request from requests, and resolve request empty array
+      _.remove(this.requests, request);
+      request.resolve([]);
+    });
+
+    this.start();
+
+    return result;
+  }
+
+  start() {
+    if (!this.started) {
+      this.started = true;
+      this.poll().catch(err => {
+        this.started = false;
+        // Resolve everything as failed
+        this.destroy();
+        let requests = this.requests;
+        this.requests = [];
+        requests.map(r => r.reject(err));
+      }).catch(err => {
+        process.nextTick(() => this.parent.emit('error', err));
+      });
+    }
+  }
+
+  async poll() {
+    // Get poll functions for pending queues (ordered by priority)
+    let polls = await this.parent._queueService.pendingQueues(
+      this.provisionerId, this.workerType,
+    );
+    // While we have requests for hints
+    while (_.sumBy(this.requests, 'count') > 0) {
+      let claimed = 0; // Count hints claimed
+
+      // In-order of priority, we poll hints from queues
+      for (let poll of polls) {
+        // While limit of hints requested is greater zero, and we are getting
+        // hints from the queue we continue to claim from this queue
+        let limit, hints;
+        let i = 10; // count iterations a limit to 10, before we start over
+        while ((limit = _.sumBy(this.requests, 'count')) > 0 &&
+               (hints = await poll(limit)).length > 0 && i-- > 0) {
+          // Count hints claimed
+          claimed += hints.length;
+
+          // While we have hints and requests for hints we resolve requests
+          while (hints.length > 0 && this.requests.length > 0) {
+            let {resolve, count} = this.requests.shift();
+            resolve(hints.splice(0, count));
+          }
+
+          // Release remaining hints (this shouldn't happen often!)
+          await Promise.all(hints.map(hint => hint.release()));
+          this.parent._monitor.count('hints-released', hints.length);
```

Comment:
how not often... report to sentry not often?

Does sentry have a system to alert if you hit an error more than X times per unit time?  Can you report positives so that you can measure percentage failure and report on that?